kornia/enhance/adjust.py

```diff
@@ -633,69 +633,116 @@ def _blend_one(input1: torch.Tensor, input2: torch.Tensor, factor: torch.Tensor)
-def _scale_channel(im: torch.Tensor) -> torch.Tensor:
-    r"""Scale the data in the channel to implement equalize.
+def _equalize_channels(
+    im: torch.Tensor,
+    bins: Optional[int] = None,
+    minimum: Optional[Union[float, int]] = None,
+    maximum: Optional[Union[float, int]] = None,
+    dtype=None,
+) -> torch.Tensor:
+    if im.is_floating_point() and bins is None:
+        raise ValueError(f"`bins` are required for floating point inputs")
+
+    # infer the bins from the dynamic range of an integer tensor
+    im_min = im.min() if minimum is None else minimum
+    im_max = im.max() if maximum is None else maximum
+    im_bins = im_max - im_min + 1 if bins is None else bins
+
+    # convert to floating point and build a histogram
+    if dtype is None and not im.is_floating_point():
+        im = im.to(torch.float)
+    elif dtype is not None:
+        im = im.to(dtype)
+
+    hist = torch.histc(im, bins=im_bins, min=im_min, max=im_max)
+
+    # start from zero and drop the last element so we have a 0.5 mean
+    hist = torch.cat([torch.zeros(1, dtype=hist.dtype, device=hist.device), hist[:-1]])
+
+    # build a lookup table normalized to [0, 1]
+    lut = torch.cumsum(hist, 0) / hist.sum()
+
+    # expand to match the input tensor dims to avoid reshaping
+    lut = lut.expand(*im.shape[:-1] + lut.shape)
+
+    # gather along the channel dimension
+    return torch.gather(lut, im.ndim - 1, im - im_min)
+
+
+def _equalize_batch(
+    im: torch.Tensor,
+    per_channel: bool = False,
+    per_image: bool = True,
+    bins: Optional[int] = None,
+    minimum: Optional[Union[float, int]] = None,
+    maximum: Optional[Union[float, int]] = None,
+    dtype=torch.float,
+) -> torch.Tensor:
+    r"""Equalize a channels last tensor to :math:`[0,1]`
 
     Args:
-        input (torch.Tensor): image tensor with shapes like :math:`(H, W)` or :math:`(D, H, W)`.
+        input (torch.Tensor): tensor with shape :math:`(*, C)` or :math:`(B, *, C)`
+        per_channel (bool): apply normalization per channel, otherwise normalize across channels
+        per_image (bool): apply normalization per image, otherwise normalize across the batch
+        bins (int): the number of bins, this will be inferred from min/max for integer tensors
+        minimum (number): values lower than the minimum will be transformed to zero
+        maximum (number): values lower than the maximum will be transformed to one
+        dtype: the datatype of the result tensor
     Returns:
-        torch.Tensor: image tensor with the batch in the zero position.
+        torch.Tensor: a tensor
     """
-    min_ = im.min()
-    max_ = im.max()
-
-    if min_.item() < 0. and not torch.isclose(min_, torch.tensor(0., dtype=min_.dtype)):
-        raise ValueError(
-            f"Values in the input tensor must greater or equal to 0.0. Found {min_.item()}."
+    if per_image:
+        return torch.stack(
+            [
+                _equalize_batch(
```

Comment:
Are you sure to call _equalize_batch here?

Reply:
yes, note the explicit recursion breaking comment. the original code does this unconditionally which causes pain for some architectures